### ISSUE TITLE
Concurrent "zed lake load" causes "malformed zng record" in journal

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#2614]
Repro is with `zed` commit a5bdb86.

I have a demo setup that's calling `zed lake load` on Zeek logs as they rotate every 60 seconds. After letting it run for a few minutes, it got into a state where I'm unable to fully read the journal because of a malformed entry.

```
$ zed lake log -p default
commit 1rX6vtBPruI8QW1soh6HIyqRXLK
Author: root@PhilMac.local
Date:   2021-04-22T16:52:00Z

    load dns.2021-04-22-09-51-32.log

    Add 1rX6vphLSLV447qSsiWFi49RXue 967B data 726B file 6 records
       from 2021-04-22T16:51:32Z to 2021-04-22T16:51:31Z

commit 1rX6vtZhBvgsw8BEVUwD9cWWDo6
Author: root@PhilMac.local
Date:   2021-04-22T16:52:00Z

    load known_hosts.2021-04-22-09-51-33.log

    Add 1rX6vpYwNfO2W2HQc0MznffNJWd 26B data 48B file 1 records
       from 2021-04-22T16:51:32Z to 2021-04-22T16:51:32Z

commit 1rX6vuDQoUZ4lMPut5BqlxPeqJy
Author: root@PhilMac.local
Date:   2021-04-22T16:52:00Z

    load stats.2021-04-22-09-51-30.log

    Add 1rX6vofrTxaOZmnKayI5zyqCC7L 59B data 417B file 1 records
       from 2021-04-22T16:51:30Z to 2021-04-22T16:51:30Z

commit 1rX6vsp4X7e6TgpU6Ey6DLmg2Qs
Author: root@PhilMac.local
Date:   2021-04-22T16:52:00Z

    load ssl.2021-04-22-09-51-33.log

    Add 1rX6vod7F9ztcziwa35zyh5h1So 110B data 374B file 1 records
       from 2021-04-22T16:51:32Z to 2021-04-22T16:51:32Z

commit 1rX6w0PbaOIicKWdk6VWetFuNAj
Author: root@PhilMac.local
Date:   2021-04-22T16:52:01Z

    load conn.2021-04-22-09-51-38.log

    Add 1rX6vwauuR03aVG1a0baa2r4pMD 788B data 834B file 9 records
       from 2021-04-22T16:51:48Z to 2021-04-22T16:51:31Z

commit 1rX73SYlExHFIioqVWjuPlvgQtH
Author: root@PhilMac.local
Date:   2021-04-22T16:53:00Z

    load dns.2021-04-22-09-52-00.log

    Add 1rX73OLtAl031vMdhNE9R8f5i1D 13.68KB data 3KiB906B file 95 records
       from 2021-04-22T16:52:52Z to 2021-04-22T16:51:43Z

malformed zng record
```

I then narrowed the problem down to the attached [6.zng.gz](https://github.com/brimdata/zed/files/6360450/6.zng.gz).

```
~/Library/Application Support/Brim/data/spaces/sp_1rX6pGXhUMIdeXfE1Pm2V2HGzQq/1rX6pI0pT900dWoLCUF2ILnvELk/L/J$ for file in *.zng; do   zq -z $file > /dev/null; done
6.zng: malformed zng record

~/Library/Application Support/Brim/data/spaces/sp_1rX6pGXhUMIdeXfE1Pm2V2HGzQq/1rX6pI0pT900dWoLCUF2ILnvELk/L/J$ zq -z 6.zng 
{commit:0x0d0f5becfd275b16ee40df063288020904a1d56b (=KSUID),segment:{id:0x0d0f5bec72e79f25f9799ea96ffd8d4456ad93a3 (KSUID),meta:{first:2021-04-22T16:52:52.688975Z,last:2021-04-22T16:51:43.468973Z,count:95 (uint64),size:13681,row_size:3978} (=Metadata)} (=Reference)} (=Add)
{commit:0x0d0f5becfd275b16ee40df063288020904a1d56b (KSUID),author:"root@PhilMac.local",date:2021-04-22T16:53:00.557311Z,message:"load dns.2021-04-22-09-52-00.log"} (=CommitMessage)
6.zng: malformed zng record
```

**Comment by mccanne** [comment]
This is a bit mysterious right now.  The zng stream ends in `67` `ff` then these last two byte are repeated causing the "corruption".  The journal files are written with ioutil.WriteFile so anything left behind by accident would be truncated.  The zng marshaler does a builder.Reset then a Bytes, so that shouldn't be creating extra bytes.  The zngio.Writer is properly closed by commit.Transaction.Serialize starting from an empty bytes.Buffer.  Given that the repeated data includes EOS, it seems like a problem in the write path after the marshaling but zngio.Writer is very old and stable.  Feels like mutex but the writer should be single threaded.  Hmm.

**Comment by philrz** [comment]
I've come up with a recipe that reproduces this pretty well. It's quite abusive of course. I'm currently repro'ing with `zed` commit e8f6d09.

If you've got the [Zeek TSV zed-sample-data](https://github.com/brimdata/zed-sample-data/tree/main/zeek-default) cloned locally, execute these commands while in the `zed-sample-data/zeek-default/` directory. It'll initialize an empty lake+pool and start populating it with repeated concurrent loads of the full set.

```
export ZED_LAKE_ROOT=$HOME/logs
rm -rf $ZED_LAKE_ROOT
zed lake init
zed lake create -p default -k ts -order desc
while true; do for file in *; do   zed lake load -p default $file & done; wait; done
```

I'm not sure if it's related, but in the output I see some other signs of unhealthiness, including this message occasionally:

```
strconv.ParseUint: parsing "": invalid syntax
```

Also this one:

```
file:///Users/phil/logs/1rXXBFSkoTKLXF0WgsVU0N7ohBE/L/J: no such journal
```

As for the `malformed zng record`, sometimes it comes right away and sometimes it takes a few minutes, but I find that if I keep executing `zed lake log -p default` from another window eventually it happens.

**Comment by mccanne** [comment]
Ah okay.  You're backgrounding the zed commands so they are running concurrently and stomping on each other's journal.  Adding transactional consistency here is part of the design but not yet implemented.  I will put in the hook for the local file system, which will solve your problem for now, but we'll also need to implement S3.

**Comment by philrz** [comment]
171  
820  
103  
364

**Comment by philrz** [comment]
Indeed, the concurrency seems to be key. I checked with @JustinAzoff from the Zeek team and he confirmed that the way I was doing it via `zeekctl`, the `archive-log` scripts (and hence the postprocessor scripts like one I had doing the `zed lake load`) are effectively all started in parallel for all the rotated logs. My `while` loop above is just a more abusive way to get the same kind of concurrency quicker.

To log a baseline before I test a fix, I just did repeated runs of the `while` loop and logged how many successful commits accumulated in the log before I saw the `malformed zng record` error. The counts across five runs were 171, 820, 103, 364, and 117. Then I started a run that _doesn't_ background them and hence does each `zed lake load` serially, and it's accumulated 6400+ successful commits and is still going without a problem, so that further validates that it's all about concurrency. I see there's a test branch up, so will try that next.

**Comment by philrz** [comment]
Verified in `zed` commit 3fb771c.

Using this fixed version of `zed`, I repeated the abusive `while` loop shown above and left it running for several minutes. Thus far I've accumulated 9000+ successful commits without a repro of the `malformed zng record`, and it's still chugging along. This is far in excess of the \<1000 commits it would make it through before triggering the problem pre-fix.

```
$ zed lake log -p default | grep commit | wc -l
    9089
```

That said, similar to what I described during the branch tests in linked PR #2616, I can see there's still some odd items appearing in the output during the abusive/concurrent loop that don't show up when I do each `zed lake load` "serially". Specifically these show up as occasional:

```
strconv.ParseUint: parsing "": invalid syntax
```

and:

```
file:///Users/phil/logs/1ragtZ8VeRNbaQaSjSUK81hOhlA/L/J: no such journal
```

I've not yet connected these to any user-visible problems with the actual data, so they may turn out to be benign. However since they're unexpected and only show up during heavy concurrency, I'll make a note-to-self in #2615 to re-run these tests after those changes land, since it looks like that will significantly overhaul how we handle concurrency and increase our testing in that area.

Thanks @mccanne!
